src/MatrixFields/field_name_dict.jl

```diff
@@ -237,6 +276,103 @@ function Base.one(matrix::FieldMatrix)
     return FieldNameDict(inferred_diagonal_keys, entries)
 end
 
+"""
+    get_field_first_index_offset(name::FieldName, ::Type{T}, ::Type{S})
+
+Returns the offset of the the field with name `name` in an object of type `S`
+in multiples of `sizeof(T)`.
+"""
+function get_field_first_index_offset(
+    name::FieldName,
+    ::Type{T},
+    ::Type{S},
+) where {T, S}
+    if name == @name()
+        return 0
+    end
+    child_name = extract_first(name)
+    child_type = fieldtype(S, child_name)
+    remaining_field_chain = drop_first(name)
+    field_index =
+        UnrolledUtilities.unrolled_findfirst(isequal(child_name), fieldnames(S))
+    return DataLayouts.fieldtypeoffset(T, S, field_index) +
+           get_field_first_index_offset(remaining_field_chain, T, child_type)
+end
+if hasfield(Method, :recursion_relation)
+    dont_limit = (args...) -> true
+    for m in methods(get_field_first_index_offset)
+        m.recursion_relation = dont_limit
+    end
+end
+
+"""
+    get_scalar_keys(dict::FieldMatrix)
+
+Returns a `FieldMatrixKeys` object that contains the keys of all the scalar
+entries in the `FieldMatrix` `dict`.
+"""
+function get_scalar_keys(dict::FieldMatrix)
+    keys_tuple = unrolled_flatmap(keys(dict).values) do key
+        entry = values(dict)[UnrolledUtilities.unrolled_findfirst(
+            isequal(key),
+            keys(dict).values,
+        )]
+        entry =
+            entry isa ColumnwiseBandMatrixField ? entry.entries.:(1) : entry
```

Comment:
This is correct for a `ColumnwiseBandMatrixField` or a `UniformScaling`, but not for a `DiagonalMatrixRow` (the other type of `ScalingFieldMatrixEntry`). When we have a single tensor in a `FieldMatrix`, we need to extract a number from that tensor for `scalar_fieldmatrix`. You'll also need to modify the relevant method for `get_internal_entry` to do this, and you'll probably want to add/modify a block in `dycore_prognostic_EDMF_FieldMatrix` to test this functionality.